src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -32,15 +41,119 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory, IDispo
+                EqualityComparer<DocumentId>.Default,
+                _listener,
+                _cancellationTokenSource.Token);
+
+            _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+            ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
+        {
+            foreach (var documentId in documentIds)
+            {
+                // Now, enqueue foreground work to actually process these documents in a serialized and incremental
+                // fashion.  FireEventsForDocument will actually limit how much time it spends firing events so that it
+                // doesn't saturate  the UI thread.
+                _notificationService.RegisterNotification(
+                    () => FireEventsForDocument(documentId),
+                    _listener.BeginAsyncOperation("CodeModelEvent"),
+                    cancellationToken);
+            }
+
+            return System.Threading.Tasks.Task.CompletedTask;
+
+            bool FireEventsForDocument(DocumentId documentId)
+            {
+                // If we've been asked to shutdown, don't bother reporting any more events.
+                if (_cancellationTokenSource.IsCancellationRequested)
+                    return false;
+
+                var projectCodeModel = this.TryGetProjectCodeModel(documentId.ProjectId);
+                if (projectCodeModel == null)
+                    return false;
+
+                var filename = _visualStudioWorkspace.GetFilePath(documentId);
+                if (filename == null)
+                    return false;
+
+                if (!projectCodeModel.TryGetCachedFileCodeModel(filename, out var fileCodeModelHandle))
+                    return false;
+
+                var codeModel = fileCodeModelHandle.Object;
+                return codeModel.FireEvents();
+            }
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            // Events that can't change existing code model items.  Can just ignore them.
+            switch (e.Kind)
+            {
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.ProjectAdded:
+                case WorkspaceChangeKind.DocumentAdded:
+                case WorkspaceChangeKind.AdditionalDocumentAdded:
+                case WorkspaceChangeKind.AdditionalDocumentRemoved:
+                case WorkspaceChangeKind.AdditionalDocumentReloaded:
+                case WorkspaceChangeKind.AdditionalDocumentChanged:
+                case WorkspaceChangeKind.AnalyzerConfigDocumentAdded:
+                case WorkspaceChangeKind.AnalyzerConfigDocumentRemoved:
+                case WorkspaceChangeKind.AnalyzerConfigDocumentReloaded:
+                case WorkspaceChangeKind.AnalyzerConfigDocumentChanged:
+                    return;
```

Comment:
>return [](start = 20, length = 6)

Would it make sense to optimize DocumentChanged events here?

Reply:
i don't know what that means.  can you clarify?

Reply:
I mean you could add these:

```
case WorkspaceChangeKind.DocumentChanged:
case WorkspaceChangeKind.DocumentRemoved:
     _documentsToFireEventsFor.AddWork(e.DocumentId);
     return;
```

---
In reply to: [401959538](https://github.com/dotnet/roslyn/pull/42965#discussion_r401959538) [](ancestors = 401959538)

Reply:
ah, i see.  smart :)